### ISSUE TITLE
Newly discovered "Behavior with BPLCON1 shifts and mid-scanline BPLCON0 change" undocumented feautere not supported

### ISSUE THREAD

**Issue by mras0** [dirkwhoffmann/vAmiga#609]
Super low priority, but may be of interest. Thread on [EAB](http://eab.abime.net/showthread.php?t=108873). You might not want to start work on it until Toni does the debugging :)

It should look like this:
![image](https://user-images.githubusercontent.com/9952017/142276443-11763ebe-5041-4462-a267-0f11da343e77.png)

(EDIT, since it's easier to compare) Looks like this in 1.1 Beta 5 (web edition):
![image](https://user-images.githubusercontent.com/9952017/142285434-90d3e407-afb3-4621-b5e9-82dd50aae507.png)

But I think it looks like this in latest vAmiga (5445aaf907a6c162eaf233584e92da9f17cfcf9a, big disclaimer follows):

![image](https://user-images.githubusercontent.com/9952017/142276552-883c89b4-df2b-4eca-b5d0-34f90e709f33.png)

Screenshot is from my very, very early attempt at getting an up to date vAmiga running locally under WSL2 (for various reasons I couldn't get it to compile easily under Windows), so it might work with correct settings.

I'm using (in addition to KS1.3):
```
        revertToFactorySettings();
        configure(CONFIG_A500_ECS_1MB);
```

But something else might be out of order, so take it with a grain of salt.

**Comment by dirkwhoffmann** [comment]
Interesting test case!

I've converted it to an ADF to make it easier to run it on real machines: [bplcon-new.adf.zip](https://github.com/dirkwhoffmann/vAmiga/files/7560231/bplcon-new.adf.zip)

On my A500 8A it looks as described on the EAB thread:
![IMG_1961](https://user-images.githubusercontent.com/12561945/142358898-cf5a2c41-f34c-4f66-bb4b-8c2333166640.JPG)

If I interpret the image correctly, I guess the part marked with the yellow arrow is the part that is "missing" in the area marked with the blue arrow. 

In the upper part, vAmiga seems to do the right thing.

**Comment by dirkwhoffmann** [comment]
What I've learned from the test case so far is that the following line in `Denise::beginOfLine` is definitely wrong:

```
    // Prepare the bitplane shift registers
    for (isize i = 0; i < 6; i++) shiftReg[i] = 0;
```

When this line is removed and the shift registers are only cleared when they were used, it looks like this: 
<img width="974" alt="Bildschirmfoto 2021-11-19 um 16 21 23" src="https://user-images.githubusercontent.com/12561945/142647527-721e5800-63c4-4436-ac39-462894ffe08c.png">

It's better, but still not right.

**Comment by dirkwhoffmann** [comment]
I probably nailed it:

<img width="974" alt="Bildschirmfoto 2021-11-19 um 17 25 55" src="https://user-images.githubusercontent.com/12561945/142659897-d1b42284-f259-4d28-b970-49ce8e9b63e6.png">

Here is a summary of what I had to change in vAmiga to make it work:  

`Denise::beginOfLine(isize vpos)`:
Don’t clear the shift registers.

`Denise::drawOdd(Pixel offset)`:
`Denise::drawEven(Pixel offset)`:
`Denise::drawBoth(Pixel offset)`:
Only clear the shift registers that have been pumped out.

`Denise::setBPLxDAT(u16 value)`:
Only feed the BLTDAT registers of enabled bitplanes into the pipeline. The BPU value is determined by the Agnus view on these bits, not the Denise view. The test program is timing-sensitive here. 

`Denise::updateShiftRegisters()`: 
Always move all pipeline registers to the shift registers, no matter how many bitplanes are enabled.

I have created a separate branch for the new code: https://github.com/dirkwhoffmann/vAMIGA/tree/issue609

Right now, the code is *not* on the dev branch, because 
- it needs to be cleaned up properly.
- I can’t assure that it’s 100% correct. It might break other stuff.

**Comment by dirkwhoffmann** [comment]
The test case written by EAB user "losso" turns out to be a very good one.

Here is a combination where vAmiga still does it wrong: 

<img width="930" alt="Bildschirmfoto 2021-11-21 um 08 03 01" src="https://user-images.githubusercontent.com/12561945/142753114-d958a1d1-d7be-4137-89c4-791f0d8e8d37.png">

It is supposed to look like this:

<img width="926" alt="Bildschirmfoto 2021-11-21 um 08 02 47" src="https://user-images.githubusercontent.com/12561945/142753090-b81f3599-5243-408b-9b1c-68d374b12174.png">

**Comment by dirkwhoffmann** [comment]
To ease debugging, I've added the ability to wipe out certain bitplanes. The functionality is accessible via the debug console:

<img width="974" alt="Bildschirmfoto 2021-11-21 um 09 21 02" src="https://user-images.githubusercontent.com/12561945/142755003-d7bb3cf3-688f-4f47-8326-75c95baadf79.png">

<img width="974" alt="Bildschirmfoto 2021-11-21 um 09 21 08" src="https://user-images.githubusercontent.com/12561945/142755005-b5a7a7f9-a938-4d65-b1a6-ef34c58d39a3.png">

The parameter is a bit mask. E.g, a value of 3 will wipe out biplane 1 and 2.

**Comment by dirkwhoffmann** [comment]
This is how the test case looks like with plane 5 disabled:

<img width="974" alt="Bildschirmfoto 2021-11-21 um 09 39 14" src="https://user-images.githubusercontent.com/12561945/142755479-2450b705-40dc-403b-b721-4fc1850ddff2.png">

Conclusion: If BPLCON1::PF1 equals 0x09, vAmiga still fills shift register 5, but the real machine does not.

**Comment by mras0** [comment]
Weird, for the $9 case, your vAmiga screenshot matches what losso posted (pf1-screenshots.zip):
![image](https://user-images.githubusercontent.com/9952017/142756197-f9746673-e41b-49a6-8f19-e8a22a8c33bd.png)

And that's also how Toni has implemented it:
![image](https://user-images.githubusercontent.com/9952017/142756246-298fdddb-dbc2-455b-8c96-04d02afa671b.png)

**Comment by dirkwhoffmann** [comment]
> Weird, for the $9 case, your vAmiga screenshot matches what losso posted (pf1-screenshots.zip

Interesting. I'll recheck this on my real machine...

**Comment by dirkwhoffmann** [comment]
I've rerun the test on my A500 and it looked exactly the same to what I've posted above. I've also checked on another A500 (the one which we call MMSE (Marble Madness Special Edition), because it became the victim of an epic retrobright fail). On this machine, it looks similar to what losso has posted:

<img width="1409" alt="Bildschirmfoto 2021-11-21 um 10 28 38" src="https://user-images.githubusercontent.com/12561945/142756757-ddd08684-b0a5-4821-89b6-8342250919b9.png">

Both are rev. 8A A500's, but I haven't checked the exact Agnus or Denise revision numbers.

**Comment by dirkwhoffmann** [comment]
@mras0: Could you do me a favour and run this combination with the latest UAE build?

<img width="1257" alt="Bildschirmfoto 2021-11-21 um 11 10 22" src="https://user-images.githubusercontent.com/12561945/142757849-6602d471-b9a0-4901-9340-db2b9c811c94.png">

**Comment by mras0** [comment]
![image](https://user-images.githubusercontent.com/9952017/142758099-e72633c9-1ec6-416d-9353-f33142b51d21.png)

Using A500 ECS Agnus quickstart. 

![image](https://user-images.githubusercontent.com/9952017/142758155-b35a152d-7235-4e25-bdd4-49d63c7dd41d.png)

Using A500 OCS (most common) quickstart.

I think they look the same, but Toni apparently found a subtle difference between ECS/AGA and OCS.

Very interesting that it can give different results on Amigas that should otherwise be very similar.

**Comment by tonioni** [comment]
Difference is between OCS Denise and ECS Denise. There are some internal differences in shifter pipeline (that some specific tests show) but more about that later...

**Comment by dirkwhoffmann** [comment]
> Difference is between OCS Denise and ECS Denise

Interesting stuff. I'm curious about the details.

I'm still struggling with a fundamental flaw in my Denise implementation. There is a specific thing which I don't understand. The following picture shows a test case using a DDF window starting at $38. The test iterates BPLCON1 through $11, $22, $33,... On my A500, the image looks as expected: 

<img width="835" alt="Bildschirmfoto 2021-11-23 um 13 27 54" src="https://user-images.githubusercontent.com/12561945/143025274-914b7158-7a24-4116-a1f9-d73e1be47ad3.png">

Now, when I change DDSTRT to $3A, it looks like this:

<img width="835" alt="Bildschirmfoto 2021-11-23 um 13 28 36" src="https://user-images.githubusercontent.com/12561945/143025411-2a8cf2bb-40c7-4346-b565-dba11d2c2842.png">

What I don't understand is how DDSTRT interacts with BPLCON1. Why does the upper part of the image jump to the right?

**Comment by tonioni** [comment]
I guess you use BPL1DAT position as BPLCON1 "reference" position. It does sound logical but it is wrong. Correct is Denise internal horizontal counter. (=if BPL1DAT position = DDFSTRT position is "unaligned" with Denise horizontal counter -> offset)

Logic something like "if ((horizontal counter & mask[7=hires,15=lores]) == BPLCON1[odd/even]) and BPLxDAT_temp has been updated (=BPLxDAT was written since last comparison match) and plane is active: copy BPLxDAT_temp to shift register. Clear "has been updated" flag.

This logic has low level differences on OCS vs ECS Denise vs AGA, for example OCS Denise seem to have parallel shifters if hires (most likely because OCS chip technology couldn't do full 14MHz), one for odd and one for even hires pixels (and most likely it is not full 2x8 but only 2x4 and copy is done twice per BPL word, to save logic gates). ECS probably was fast enough to support hires with single shifter: lores and hires were most likely merged, shres most likely uses dual pipeline. AGA probably got rid of dual stuff completely. This is speculation only but at least OCS Denise "dual" behavior is quite obvious, have lores single bitplane with all bits set to one, switch to hires mid-scanline and you'll notice interesting 2x4 pixel pattern (make sure BPLCON1=0 to make it very clear). This pattern does not appear if ECS Denise or AGA.

**Comment by dirkwhoffmann** [comment]
@tonioni Thanks a lot for the explanation. I get the idea, but I am still not sure if I understand all pieces. In the following, I distinguish four events:

    aa : Read BPL1DAT from memory (DMA)
    bb : Transfer BPL1DAT, ..., BPL5DAT to temp registers
    cc : Transfer temp registers to shift registers (if bb happened before)
    dd : Start shift registers

Let's examine the following four cases:

1. DDF is aligned, no bitplane shift
2. DDF is aligned, bitplane shift
3. DDF is misaligned, no bitplane shift
4. DDF is misaligned, bitplane shift

In my current understanding, the following happens. Is this correct or am I still missing something? 

```
1. DDFSTRT = 38
   BPLCON1 = 00

   <---- Fetch unit -----> <---- Fetch unit -----> <---- Fetch unit ----->
37 38 39 3A 3B 3C 3D 3E 3F 40 41 42 43 44 45 46 47 48 49 4A 4B 4C 4D 4E 4F
                        aa                      aa                      aa
                        bb                      bb                      bb
   cc          cc          cc          cc          cc          cc
                           dd                      dd 

2. DDFSTRT = 38
   BPLCON1 = 44

   <---- Fetch unit -----> <---- Fetch unit -----> <---- Fetch unit ----->
37 38 39 3A 3B 3C 3D 3E 3F 40 41 42 43 44 45 46 47 48 49 4A 4B 4C 4D 4E 4F
                        aa                      aa                      aa
                        bb                      bb                      bb
   cc          cc          cc          cc          cc          cc
                                 dd                      dd 


3. DDFSTRT = 3A
   BPLCON1 = 00

   <---- Fetch unit -----> <---- Fetch unit -----> <---- Fetch unit ----->
37 38 39 3A 3B 3C 3D 3E 3F 40 41 42 43 44 45 46 47 48 49 4A 4B 4C 4D 4E 4F
                        aa                      aa                      aa
                        bb                      bb                      bb
         cc          cc          cc          cc          cc          cc
                           dd                      dd 


4. DDFSTRT = 3A
   BPLCON1 = 44

   <---- Fetch unit -----> <---- Fetch unit -----> <---- Fetch unit ----->
37 38 39 3A 3B 3C 3D 3E 3F 40 41 42 43 44 45 46 47 48 49 4A 4B 4C 4D 4E 4F
                        aa                      aa                      aa
                        bb                      bb                      bb
         cc          cc          cc          cc          cc          cc
                                 dd                      dd 

```

**Comment by tonioni** [comment]
Assuming ECS Agnus (=DDFSTRT 3A is valid, bit 1 exists in DDFSTRT). Fetch start can be unaligned, Agnus does not care, if DDFSTRT is 3A, it will start from 3A.

**Comment by dirkwhoffmann** [comment]
After working on remote servers for a while, I am back at improving emulation accuracy. With my latest code changes, I think I got it right. All value combinations that didn't work previously seem to work fine now. I.e., combination C9 / 5 gives the correct result now:

<img width="976" alt="Bildschirmfoto 2022-01-03 um 17 31 41" src="https://user-images.githubusercontent.com/12561945/147968772-57fd1fb3-48c6-4b22-a3c8-dec94a2dbd18.png">

**Comment by dirkwhoffmann** [comment]
> Super low priority, but may be of interest.

@mras0: Thanks again for starting this thread. Losso's test case revealed a big bug in vAmiga. In hindsight the priority was huge (which was impossible to know at the beginning of course).

**Comment by mras0** [comment]
No problemo, always happy to create extra work for others ;)

When I have some time, I'll update the vAmigaSDL port to the latest version, and maybe take a stab at adding recorder support for Windows if you're open to incorporating a patch/pull request.

**Comment by dirkwhoffmann** [comment]
> take a stab at adding recorder support for Windows if you're open to incorporating a patch/pull request.

Yes, cool. I'd be happy to merge pull requests in.  

> you should just close the issue when you feel like it,

You mean this issue? When an issue is solved, I usually put a label on it (V1.1 in this case) and close all labeled issues altogether when a new version is releases. This makes it easy for me to remember all changes to the previous release (I am usually too lazy to keep separate release notes).

**Comment by mras0** [comment]
> > take a stab at adding recorder support for Windows if you're open to incorporating a patch/pull request.
> 
> Yes, cool. I'd be happy to merge pull requests in.

👍 

> 
> > you should just close the issue when you feel like it,
> 
> You mean this issue? When an issue is solved, I usually put a label on it (V1.1 in this case) and close all labeled issues altogether when a new version is releases. This makes it easy for me to remember all changes to the previous release (I am usually too lazy to keep separate release notes).

Great, yes, I meant this issue. Was just reminded by the "close with comment" button that I was the reporter :)

**Comment by dirkwhoffmann** [comment]
Fixed in v1.1b8
